pkg/util/pubsub/http_handler.go

```diff
@@ -0,0 +1,160 @@
+package pubsub
+
+import (
+	"context"
+	"net/http"
+	"time"
+
+	"github.com/gomodule/redigo/redis"
+	"nhooyr.io/websocket"
+
+	"github.com/authgear/authgear-server/pkg/util/log"
+)
+
+type RedisPool interface {
+	Get() redis.Conn
+}
+
+type WebsocketOutgoingMessage struct {
+	MessageType websocket.MessageType
+	Data        []byte
+}
+
+type Delegate interface {
+	Accept(r *http.Request) (channelName string, err error)
+	OnWebsocketMessage(messageType websocket.MessageType, data []byte) (*WebsocketOutgoingMessage, error)
+}
+
+const (
+	WebsocketReadTimeout  = 30 * time.Second
+	WebsocketWriteTimeout = 10 * time.Second
+	RedisReadTimeout      = 30 * time.Second
+)
+
+// HTTPHandler receives incoming websocket messages and delegates them to the delegate.
+// For each websocket connection, a redis pubsub connection is established.
+// Every message from the redis pubsub connection is forwarded to the websocket connection verbatim.
+type HTTPHandler struct {
+	RedisPool     RedisPool
+	Delegate      Delegate
+	LoggerFactory *log.Factory
+}
+
+func (h *HTTPHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
```

Comment:
The data flow of this handler seems a bit unnatural:
- Redis routine: Redis read -> WebSocket write
- WebSocket routine: WebSocket read -> delegate (?) -> WebSocket write (?)

It would be unclear to API consumer on what delegate must perform in order to complete the data flow. If I understand correctly:
- it publish the incoming message to Redis
- return some message (?) to be written to WebSocket.

I'd suggest to either abstract away Redis subscription in delegate, or handle Redis publish directly in the handler.

Reply:
Actually we do not expect the websocket client to send us any message at the moment. For simplicity, maybe we can remove `OnWebsocketMessage` from `Delegate`. This also makes sure `wsConn.Write` is only called within the websocket goroutine.